### DISCUSSION TITLE
vercel/og gives different font weight results in playground and Next project

### DISCUSSION THREAD

**Discussion by hypervillain** [vercel/community#1343] [Help] [answered]
### Summary

I've had a lot of fun playing around with `vercel/og` recently, but I'm unable to reproduce the OG playground results in my own Next project. I'm unable to set font weights in `tw` strings (eg. `font-bold` with Tailwind) nor `fontWeight` in style prop. Updating these values works correctly in the playground though. I've tried adding custom fonts without success: the font still looks like the default (Satori ?) one.

You can view the expected result [in the playground](https://og-playground.vercel.app/?share=hVVbT9swFP4rR94DL3Ub2lEggj6M9QFtGhOgTZP64iZOYvAls50mXdX_vuOU0lICVJWTc_vO3VmRxKScxOQiFYuZBvw5v5T8crXaUAAFF3nhYzg6jqJFcdTb8muR-uKJXe_YqXClZEsUZJI3O36gvgrLEy-MRmliZKX0Ts6kyPW158oFIdee253woXJeZMsrg2wdYjlUmLPkMbem0umVkcaiRl0Iz3cKkKHtnfjHUdQ_Oz2xXG2F63V4TsIRqgC-vpwRzxtPc8uW9DyKgJUlJAjAhOYWVENZ5Q04FZcN_QzlEo-QX3tQzGxGWrh9QMUaWtNRI2e-tMFIhuP5TW0gu1H2ceY5bVODNlue0hGUdPym4b7p2zr7WpjN6B24DsjEKIX9eKXWqrpFvlHdxptVUkJNj4dQ4DEj0CipHSoU3pfxYFDXdb8e9Y3NB8MoigYIgEoLwesvpkG1CCI4C_9Of63PMGWb-UT9oPg0xFsqE1Li-6eMj8fnHHEuBsFkgrYdGYQIXnl6WQKcFIMpgQqtfL92B1MxR6ugG4JgUFiehcAwxg02zhydG5mCFPqRpsw-guL0GLXvfJUKA9-Mxv_FgE2w1CXT3RU5GOlTHOl2iNAi4VQbWltWhugD8hBStnTAcoOpIyRWB-P9KI3hNo3vBncLROkqBWlYRnDCA1Pc96Dy8LcS7cCYFBeIayfmleuBSFEmNEND3DAlXFceuIF4DfiwJwuh-nCtIVi0TtCZCFcCdy44KnFNM4Gk9lAanLugwFkfpgYzQ6mwostFxqtcMA8pVyax6EuB5v2PC1A4jzcQYEmVo5vLCfaL0uUL-33QlzH2paUaF2b2eRgmv66nv-HmB_y5D53-EMzimu1hhcaOXgLeTn_e3L4B1p1tF_c175Dzkt6ndu_bt-2T9Igpw1fCkXhF2jUm8flZ1CObLSbxyelJj6R8XuUkzph0vEewXw_iflmGb5mvWwpxwv5M1ZynJPa24uv1fw). And the current result I get can be seen here: https://comments-5tqo.vercel.app/api/og.

My edge function is very standard:

```js
return new ImageResponse(
    (
      <Comment
        scale
        classProp="tw"
        defaultFontSize={17}
        fontFamilies={{ normal: 'RobotoRegular', bold: 'RobotoBold' }}
      />
    ),
    {
      width: 2400,
      height: 1260,
      fonts: [
        {
          name: 'Roboto',
          data: roboto,
          weight: 400,
          style: 'normal',
        },
        {
          name: 'RobotoBold',
          data: robotoBold,
          weight: 700,
          style: 'normal',
        },
      ],
    },
  );
````

And the `Comment` component is nothing unusual except that I use it both to render in the website and in the ImageResponse (`withClassProp("tw" | "className"))`:

```jsx
<a
    href="#"
    {...withClassProp(classProp, "font-bold")}
    style={{
        fontSize: '16px',
        fontWeight: '100', // no effect in ImageResponse
        ...fontFamilies ? { fontFamily: `"${fontFamilies.bold}"`, } : null // no effect in ImageResponse
     }} 
>
   Studio KonKon
</a>
````

Do you have any clue? Thanks!

### Example

https://comments-5tqo.vercel.app/api/og

**Comment by ccrsxx** [comment, marked as answer]
I found the solution. 

https://github.com/vercel/satori/issues/263#issuecomment-1500705357

So by default, quoted from the above comment by the @vercel-og maintainer. 
> the @vercel/og package has only 1 built-in font (Noto Sans) with the 400 weight.

You need to include your own custom font with the weight you want. For example, if I want to have a regular and bold font, then I need to add the two fonts with different weights to the image response.

Example code:
```tsx
import { ImageResponse } from '@vercel/og';
import type { PageConfig } from 'next';
import type { NextRequest } from 'next/server';

const regularFont = fetch(
  new URL('/public/assets/inter-regular.ttf', import.meta.url)
).then((res) => res.arrayBuffer());

const boldFont = fetch(
  new URL('/public/assets/inter-bold.ttf', import.meta.url)
).then((res) => res.arrayBuffer());

export default async function handler(
  req: NextRequest
): Promise<ImageResponse> {
  const [regularFontData, boldFontData] = await Promise.all([
    regularFont,
    boldFont
  ]);

  return new ImageResponse(
    (
      <div
        style={{
          height: '100%',
          width: '100%',
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'center',
          backgroundColor: '#fff',
          fontSize: 32
        }}
      >
        <svg
          width='75'
          viewBox='0 0 75 65'
          fill='#000'
          style={{ margin: '0 75px' }}
        >
          <path d='M37.59.25l36.95 64H.64l36.95-64z'></path>
        </svg>
        <div style={{ marginTop: 40, fontWeight: 400 }}>Regular</div>
        <div style={{ marginTop: 40, fontWeight: 700 }}>Bold</div>
      </div>
    ),
    {
      width: 1200,
      height: 600,
      fonts: [
        {
          name: 'Inter',
          data: regularFontData,
          weight: 400
        },
        {
          name: 'Inter',
          data: boldFontData,
          weight: 700
        }
      ]
    }
  );
}

export const config: PageConfig = {
  runtime: 'edge'
};
```

**Comment by RustLoverTheCoder** [comment]
> ### Summary
> I've had a lot of fun playing around with `vercel/og` recently, but I'm unable to reproduce the OG playground results in my own Next project. I'm unable to set font weights in `tw` strings (eg. `font-bold` with Tailwind) nor `fontWeight` in style prop. Updating these values works correctly in the playground though. I've tried adding custom fonts without success: the font still looks like the default (Satori ?) one.
> 
> You can view the expected result [in the playground](https://og-playground.vercel.app/?share=hVVbT9swFP4rR94DL3Ub2lEggj6M9QFtGhOgTZP64iZOYvAls50mXdX_vuOU0lICVJWTc_vO3VmRxKScxOQiFYuZBvw5v5T8crXaUAAFF3nhYzg6jqJFcdTb8muR-uKJXe_YqXClZEsUZJI3O36gvgrLEy-MRmliZKX0Ts6kyPW158oFIdee253woXJeZMsrg2wdYjlUmLPkMbem0umVkcaiRl0Iz3cKkKHtnfjHUdQ_Oz2xXG2F63V4TsIRqgC-vpwRzxtPc8uW9DyKgJUlJAjAhOYWVENZ5Q04FZcN_QzlEo-QX3tQzGxGWrh9QMUaWtNRI2e-tMFIhuP5TW0gu1H2ceY5bVODNlue0hGUdPym4b7p2zr7WpjN6B24DsjEKIX9eKXWqrpFvlHdxptVUkJNj4dQ4DEj0CipHSoU3pfxYFDXdb8e9Y3NB8MoigYIgEoLwesvpkG1CCI4C_9Of63PMGWb-UT9oPg0xFsqE1Li-6eMj8fnHHEuBsFkgrYdGYQIXnl6WQKcFIMpgQqtfL92B1MxR6ugG4JgUFiehcAwxg02zhydG5mCFPqRpsw-guL0GLXvfJUKA9-Mxv_FgE2w1CXT3RU5GOlTHOl2iNAi4VQbWltWhugD8hBStnTAcoOpIyRWB-P9KI3hNo3vBncLROkqBWlYRnDCA1Pc96Dy8LcS7cCYFBeIayfmleuBSFEmNEND3DAlXFceuIF4DfiwJwuh-nCtIVi0TtCZCFcCdy44KnFNM4Gk9lAanLugwFkfpgYzQ6mwostFxqtcMA8pVyax6EuB5v2PC1A4jzcQYEmVo5vLCfaL0uUL-33QlzH2paUaF2b2eRgmv66nv-HmB_y5D53-EMzimu1hhcaOXgLeTn_e3L4B1p1tF_c175Dzkt6ndu_bt-2T9Igpw1fCkXhF2jUm8flZ1CObLSbxyelJj6R8XuUkzph0vEewXw_iflmGb5mvWwpxwv5M1ZynJPa24uv1fw). And the current result I get can be seen here: https://comments-5tqo.vercel.app/api/og.
> 
> My edge function is very standard:
> 
> ```js
> return new ImageResponse(
>     (
>       <Comment
>         scale
>         classProp="tw"
>         defaultFontSize={17}
>         fontFamilies={{ normal: 'RobotoRegular', bold: 'RobotoBold' }}
>       />
>     ),
>     {
>       width: 2400,
>       height: 1260,
>       fonts: [
>         {
>           name: 'Roboto',
>           data: roboto,
>           weight: 400,
>           style: 'normal',
>         },
>         {
>           name: 'RobotoBold',
>           data: robotoBold,
>           weight: 700,
>           style: 'normal',
>         },
>       ],
>     },
>   );
> ```
> 
> And the `Comment` component is nothing unusual except that I use it both to render in the website and in the ImageResponse (`withClassProp("tw" | "className"))`:
> 
> ```js
> <a
>     href="#"
>     {...withClassProp(classProp, "font-bold")}
>     style={{
>         fontSize: '16px',
>         fontWeight: '100', // no effect in ImageResponse
>         ...fontFamilies ? { fontFamily: `"${fontFamilies.bold}"`, } : null // no effect in ImageResponse
>      }} 
> >
>    Studio KonKon
> </a>
> ```
> 
> Do you have any clue? Thanks!
> 
> ### Additional information
> _No response_
> 
> ### Example
> https://comments-5tqo.vercel.app/api/og

what code editor do you use？
